server/src/test/java/org/elasticsearch/script/TimeSeriesCounterTests.java

```diff
@@ -0,0 +1,479 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script;
+
+import org.elasticsearch.test.ESTestCase;
+import org.hamcrest.Matcher;
+import org.junit.Before;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.script.TimeSeriesCounter.Counter;
+import static org.elasticsearch.script.TimeSeriesCounter.HOUR;
+import static org.hamcrest.Matchers.lessThan;
+
+public class TimeSeriesCounterTests extends ESTestCase {
+    protected long now;
+    protected long customCounterResolution;
+    protected long customCounterDuration;
+    protected TimeSeriesCounter tsc = new TimeSeriesCounter();
+    protected final Matcher<Long> fiveDelta = lessThan(tsc.fiveMinutes.resolution);
+    protected final Matcher<Long> fifteenDelta = lessThan(tsc.fifteenMinutes.resolution);
+    protected final Matcher<Long> twentyFourDelta = lessThan(tsc.twentyFourHours.resolution);
+    protected List<Long> events;
+    protected Counter counter;
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        now = 1635182590;
+        customCounterResolution = 45;
+        customCounterDuration = 900;
+        reset();
+    }
+
+    protected void reset() {
+        tsc = new TimeSeriesCounter();
+        events = new ArrayList<>();
+        counter = new Counter(customCounterResolution, customCounterDuration);
+    }
+
+    public void testCounterNegativeResolution() {
+        IllegalArgumentException iae = expectThrows(IllegalArgumentException.class, () -> new Counter(-20, 200));
+        assertEquals("resolution [-20] must be greater than zero", iae.getMessage());
+    }
+
+    public void testCounterNegativeDuration() {
+        IllegalArgumentException iae = expectThrows(IllegalArgumentException.class, () -> new Counter(20, -200));
+        assertEquals("duration [-200] must be greater than zero", iae.getMessage());
+    }
+
+    public void testCounterIndivisibleResolution() {
+        IllegalArgumentException iae = expectThrows(IllegalArgumentException.class, () -> new Counter(3, 101));
+        assertEquals("duration [101] must divisible by resolution [3]", iae.getMessage());
+    }
+
+    public void testOnePerSecond() {
+        long time = now;
+        long t;
+        long next = randomLongBetween(1, HOUR);
+        long twentyFive = 25 * HOUR;
+        for (int i = 0; i < twentyFive; i++) {
+            t = time + i;
+            inc(t);
+
+            if (i == next) {
+                TimeSeries ts = tsc.timeSeries(t);
+                assertThat(five(t) - ts.fiveMinutes, fiveDelta);
```

Comment:
On bucket rollover, we lose the last `resolution` of history. Say we have two buckets of resolution 5 each and increments at 0,1,3,4,5,6,7,8,9,10.  Before the increment at 10, the state will be:
```
     [0]         [1]
|_____4_____|_____5_____|
0->      <-4|5->     <-9|
```
Because 0,1,3,4 are in bucket 0 and 5,6,7,8,9 are in bucket 1.

At 10, the state will be
```
     [0]         [1]
|_____1_____|_____5_____|
10->    <-14|5->     <-9|
```
By incrementing at 10, we've wiped out the history of t=0 -> t=4.

Because this test increments one per second, we know that at most resolution will be missed due to this roll over.

Options:
1) Clarify in comments that the Counter covers between `duration - resolution + 1`  and `duration` seconds.  In the example above, we have history of the 5,6,7,8,9,10 increments which a range of 6 seconds.
2) Hold the counter for the current bucket outside of the array and do not wipe the contents of the current bucket on rollover.  This means `Counter` will cover between `duration` and `duration + resolution` seconds depending on when it's inspected.